pykokkos/lib/ufuncs.py

```diff
@@ -1381,6 +1389,270 @@ def exp2(view):
+    out[tid] = viewA[tid] / viewB[0]
+
+
+@pk.workunit
+def divide_num_2d_impl_1d_double(tid: int, viewA: pk.View2D[pk.double], viewB: pk.View1D[pk.double], out: pk.View2D[pk.double]):
+    for i in range(viewA.extent(1)):
+        out[tid][i] = viewA[tid][i] / viewB[i]
+
+
+def divide_num(viewA, viewB):
+    if (not hasattr(viewB, '__iter__')):
+        view_temp = pk.View([1], pk.double)
+        view_temp[0] = viewB
+        viewB = view_temp
+
+    if len(viewA.shape) == 2:
+        out = pk.View(viewA.shape, pk.double)
+        pk.parallel_for(
+            viewA.shape[0],
+            divide_num_2d_impl_1d_double,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+    elif str(viewA.dtype) == "DataType.double":
+        out = pk.View([viewA.shape[0]], pk.double)
+        pk.parallel_for(
+            viewA.shape[0],
+            divide_num2_impl_1d_double,
+            viewA=viewA,
+            viewB=view_temp,
+            out=out)
+    else:
+        raise RuntimeError("Incompatible Types")
+    return out
+
+@pk.workunit
+def mul0d_impl_1d_double(tid: int, viewA: pk.View1D[pk.double], viewB: pk.View1D[pk.double], out: pk.View1D[pk.double]):
+    out[tid] = viewA[tid] * viewB[0]
+
+
+def mul_num(viewA, viewB):
+    view_temp = pk.View([1], pk.double)
+    view_temp[0] = viewB
+
+    viewc = pk.View(viewA.shape, pk.double)  # 
+    viewc[:] = viewA
+
+    if str(viewc.dtype) in {"DataType.double", "DataType.float64"}:
+        out = pk.View([viewc.shape[0]], pk.double)
+        pk.parallel_for(
+            viewA.shape[0],
+            mul0d_impl_1d_double,
+            viewA=viewc,
+            viewB=view_temp,
+            out=out)
+    else:
+        raise RuntimeError("Incompatible Types")
+    return out
+
+
+@pk.workunit
+def add_numm_impl_1d_double(tid: int, viewA: pk.View1D[pk.double], viewB: pk.View1D[pk.double], out: pk.View1D[pk.double]):
+    out[tid] = viewA[tid] + viewB[0]
+
+@pk.workunit
+def add_num_2d_impl_1d_double(tid: int, viewA: pk.View2D[pk.double], viewB: pk.View1D[pk.double], out: pk.View2D[pk.double]):
+    for i in range(viewA.extent(1)):
+        out[tid][i] = viewA[tid][i] + viewB[i]
+
+def add_num(viewA, viewB):
+    if (not hasattr(viewB, '__iter__')):
+        view_temp = pk.View([1], pk.double)
+        view_temp[0] = viewB
+        viewB = view_temp
+
+    if len(viewA.shape) == 2:
+        out = pk.View(viewA.shape, pk.double)
+        pk.parallel_for(
+            viewA.shape[0],
+            add_num_2d_impl_1d_double,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+    elif str(viewA.dtype) in "DataType.double":
+        out = pk.View([viewA.shape[0]], pk.double)
+        pk.parallel_for(
+            viewA.shape[0],
+            add_numm_impl_1d_double,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+    else:
+        raise RuntimeError("Incompatible Types")
+    return out
+
+def _get_shape(arr):
```

Comment:
This looks like it is patching an oversight somewhere if you really need it? Can you summarize why our current size/shape infrastructure wasn't cutting if for you? I though we were doing pretty well in that regard apart from 0-d arrays maybe.

I'm also slightly confused by the design decision to return list objects when Nader just adjusted us to comply with the array API shape requirement of being tuple-like in gh-78.

Reply:
Removed and moved all the lists to pk.view types. We dont need this function anymore.